src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -39,3 +51,64 @@ def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
         )
 
     return QueryGenerator(plan.session, snowflake_create_table_plan_info)
+
+
+def replace_child(
+    parent: PlanTreeNode, old_node: PlanTreeNode, new_node: Selectable
+) -> None:
+    """
+    Helper function to replace the child node in the plan.
+    """
+    if not parent._is_valid_for_replacement:
+        raise ValueError(f"parent node {parent} is not valid for replacement.")
+
+    if old_node not in getattr(parent, "children_plan_nodes", parent.children):
+        raise ValueError(f"old_node {old_node} is not a child of parent {parent}.")
+
+    if isinstance(parent, SnowflakePlan):
+        assert parent.source_plan is not None
+        replace_child(parent.source_plan, old_node, new_node)
+        return
+
+    if isinstance(parent, SelectStatement):
+        parent.from_ = new_node
+        return
+
+    if isinstance(parent, SetStatement):
+        parent._nodes = [
+            node if node != old_node else new_node for node in parent._nodes
+        ]
+        return
+
+    if isinstance(parent, Selectable):
+        assert parent.snowflake_plan.source_plan is not None
+        replace_child(parent.snowflake_plan.source_plan, old_node, new_node)
+        return
+
+    if isinstance(parent, (UnaryNode, Limit, CopyIntoLocationNode)):
+        parent.children = [new_node]
+        parent.child = new_node
+        return
+
+    if isinstance(parent, BinaryNode):
+        parent.children = [
+            node if node != old_node else new_node for node in parent.children
+        ]
+        if parent.left == old_node:
+            parent.left = new_node
+        else:
```

Comment:
the left and right can be the one to be replaced if they happen to be the same node, so we can not do if else here